numcodecs/compat.py

```diff
@@ -48,8 +49,8 @@ def ensure_ndarray(buf):
 
     """
 
-    if isinstance(buf, np.ndarray):
-        # already a numpy array
+    if isinstance(buf, (np.ndarray, cp.ndarray)):
+        # already an array
         arr = buf
```

Comment:
This is easy enough to do, but it would be nice if we didn't need to learn about CuPy here.

We will probably need to find another solution for Numcodecs in the near term, but wanted to mention some related things. Namely there is some discussion about generally checking array types in issue ( https://github.com/numpy/numpy/issues/14891 ).

Also mentioned by Matt in the aforementioned issue, Dask has [`is_arraylike`]( https://github.com/dask/dask/blob/31af7f7040643c447a72c87a8f12457094ec15ff/dask/utils.py#L1079-L1101 ), which presents a way of solving this problem. I'm not sure it totally helps us here as we really need to check that there is a data buffer we can grab and use, but it is at least a good starting point.

Another thought I had is we could registered `ndarray` types and their conversion mechanisms with Numcodecs. This could allow users to add their own array types (like CuPy), but not complicate things too much here.

Other thoughts on this would generally be welcome. 🙂

Reply:
> as we really need to check that there is a data buffer we can grab and use, but it is at least a good starting point.

Perhaps one could check for the presence of a `__array__` or `__cuda_array_interface__` protocol?

Reply:
That's a good point. Thanks! Will give that a look.

Reply:
Yep, if goal here is to ultimately enable codecs with GPU implementations then using ``__cuda_array_interface__`` sounds good. E.g., when using numba cuda functions you get back an instance of some other class (DeviceArray or something like that IIRC) but that exports ``__cuda_array_interface__`` too.